test/common/stats/stat_merger_fuzz_test.cc

```diff
@@ -0,0 +1,65 @@
+#include <algorithm>
+
+#include "common/stats/stat_merger.h"
+
+#include "test/common/stats/stat_test_utility.h"
+#include "test/fuzz/fuzz_runner.h"
+#include "test/fuzz/utility.h"
+
+namespace Envoy {
+namespace Stats {
+namespace Fuzz {
+
+void testDynamicEncoding(absl::string_view data, SymbolTable& symbol_table) {
+  StatNameDynamicPool dynamic_pool(symbol_table);
+  StatNamePool symbolic_pool(symbol_table);
+  std::vector<StatName> stat_names;
+
+  for (uint32_t index = 0; index < data.size();) {
+    // Select component lengths between 0 and 7 bytes inclusive, and ensure it
+    // doesn't overrun our buffer. It's OK to get very small or empty segments.
+    uint32_t num_bytes = data[index] & 0x7;
+    num_bytes = std::min(static_cast<uint32_t>(data.size() - 1),
+                         num_bytes); // restrict number up to the size of data
+
+    // Carve out the segment and use the 4th bit from the control-byte to
+    // determine whether to treat this segment symbolic or not.
+    absl::string_view segment = data.substr(index, num_bytes);
+    bool is_symbolic = (data[index] & 0x8) == 0x0;
+    index += num_bytes + 1;
+    if (is_symbolic) {
+      stat_names.push_back(symbolic_pool.add(segment));
+    } else {
+      stat_names.push_back(dynamic_pool.add(segment));
+    }
+  }
+
+  SymbolTable::StoragePtr joined = symbol_table.join(stat_names);
+  StatName stat_name(joined.get());
+
+  StatMerger::DynamicContext dynamic_context(symbol_table);
+  std::string name = symbol_table.toString(stat_name);
+  StatMerger::DynamicsMap dynamic_map;
+  dynamic_map[name] = StatMerger::DynamicContext::encodeComponents(stat_name);
+
+  StatName decoded = dynamic_context.makeDynamicStatName(name, dynamic_map);
+  FUZZ_ASSERT(name == symbol_table.toString(decoded));
+  FUZZ_ASSERT(stat_name == decoded);
+}
+
+// Fuzzer for symbol tables.
+DEFINE_FUZZER(const uint8_t* buf, size_t len) {
+  FuzzedDataProvider provider(buf, len);
+  FakeSymbolTableImpl fake_symbol_table;
+  SymbolTableImpl symbol_table;
+
+  while (provider.remaining_bytes() != 0) {
```

Comment:
Why run this loop multiple times instead of passing the whole bytestring into `testDynamicEncoding` and using a `FuzzedDataProvider` inside the method to `ConsumeBytes(7)` for the segments?

Reply:
This is a good question. I was thinking that the len passed into this function may get very long, and I didn't think it was that useful in fuzzing this code to make huge long strings; I thought it would be more useful to have large #s of bounded-size strings. WDYT?

Reply:
Hmm `ConsumeRandomLengthString` may return strings up to size `len` so the strings could be very large here. So I don't see a difference between running this loop once or many times, since the heuristics of what the provider will do are completely unknown.

One of the fuzzing engine uses the corpus's largest entry to bound the length of the inputs, but OSS-Fuzz has more than just one engine, some of which don't constrain the input at all. They recommend that if you need strict bounds (but I don't think that's the case), you enforce the size in the fuzzer.

Reply:
makes sense; done.